Cody Tseng Is Building a Social App That Follows Your Heart, Not an Algorithm (#791)
OpenSats/website: data/blog/developer-spotlight-codytseng-jumblesocial.mdx

---

> title: "Cody Tseng Built a Social App on Nostr That Follows Your Heart, Not an Algorithm"

Reviewer: ```suggestion
title: "Cody Tseng Built a Social App That Follows Your Heart, Not an Algorithm"
```

I'd remove the "on Nostr" part.

versions][jumble-ouro]. He browses those versions himself, mining them for ideas
he hadn't thought of. 

"I'm a person with no much planning and often change my mind randomly," he says.
The only fixed point is a line in his GitHub profile: four characters in
Chinese, 随心而动. Follow your heart.

The project is called [Jumble][jumble-social]. Open it in a browser and you see
what you might see anywhere: posts, photos, and strangers saying GM — good
morning — to the internet, and other mumble jumble. What makes it different is
that nothing has been curated before you arrive, and posts sit without scores
beneath them. Jumble refuses the usual sugar rush of the social web. It asks a
question most apps have long since answered for you: what do you want to read?

<Image
  src="/static/images/spotlight/cody-tseng/jumble-screenshot.png"
  darkSrc="/static/images/spotlight/cody-tseng/jumble-screenshot-dark.png"
  alt="The Jumble web client"
  width={1920}
  height={970}
  className="w-full h-auto"
/>

Most social media is a building someone else owns. They built the walls, set the
rules, and kept the keys. Your account, posts, and followers all live inside

> it.Every profile looks the same, arranged identically, offering the same

Author: ```suggestion
it. Every profile looks the same, arranged identically, offering the same
```

his code, running inside a project used by strangers. That feeling is, he says,
what has driven him ever since.

Like most children in China, his life was structured around a single variable:
examinations. School, exams, more school, more exams, all the way to university,
and none of it suited him. "I was never the kind of student who followed rules
very well," he says. "I disliked exams and memorizing things that felt
meaningless to me." The consequence was a university he hadn't aimed for and a
major that wasn't computer science. Rules that felt arbitrary always chafed him.
So he follows his own code.

Outside of those structures, he moves at a different pace. On weekends he walks
with his girlfriend, to the sea, through parks, into bookstores, to restaurants
they haven't tried before. He doesn't plan far ahead. The four characters on his
GitHub profile are, he says, a reminder not to overthink, to act on what feels
right inside.

He doesn't remember exactly when he first encountered Bitcoin, only what made
him stay.

"Like many people, I dislike being controlled and want to truly own certain
things. I think many Nostr users feel similarly. They want to truly own their

> data instead of having it locked inside the databases of large companies." 

Author: ```suggestion
> Like many people, I dislike being controlled and want to truly own certain
> things. I think many Nostr users feel similarly. They want to truly own their
> data instead of having it locked inside the databases of large companies. 
```

Reviewer: If we would want to do that we'd use CSS styling.

that have been noticed, which means that, for now, obscurity is its own form of
protection. "Once you are using a VPN," he says, "where you are physically
located starts to matter a lot less." He laughs, as if the obstacle is real but
not where the real difficulty lies.

> Personally, I think improving the Nostr ecosystem itself is ultimately more
> important. Network access problems can usually be worked around one way or
> another, but whether the ecosystem is actually useful and interesting matters
> more.

Cody admits that if he wasn’t developing Jumble, he’d probably not even use
Nostr that much. The content on X, Reddit, RedNote, and elsewhere is simply
better. But that leaves Nostr with a chicken-and-egg problem: people need a
reason to come before there is much to read, and there will not be much to read
until more people come. To grow beyond builders and believers, it must offer
reasons to stay that have little to do with private keys or digital ownership.
People go where other people already are. 

> My vision is that in the future, when people talk about Nostr social
> applications, they will naturally think of Jumble.

> He pauses.

Author: ```suggestion
But.
```

Author: He doesnt really pause; I needed to change that. Another way to say it is: 

The second answer came differently.

---
title: "Cody Tseng Built a Social App That Follows Your Heart, Not an Algorithm"

> date: '2026-05-27'

Author: ```suggestion
date: '2026-06-01'
```

that nothing has been curated before you arrive, and posts sit without scores
beneath them. Jumble refuses the usual sugar rush of the social web. It asks a
question most apps have long since answered for you: what do you want to read?

<Image
  src="/static/images/spotlight/cody-tseng/jumble-screenshot.png"
  darkSrc="/static/images/spotlight/cody-tseng/jumble-screenshot-dark.png"
  alt="The Jumble web client"
  width={1920}
  height={970}
  className="w-full h-auto"
/>

Most social media is a building someone else owns. They built the walls, set the
rules, and kept the keys. Your account, posts, and followers all live inside
it. Every profile looks the same, arranged identically, offering the same
options. Boring, suburban gated communities, the whole lot of them. If they
change the terms, it's their way or the highway.

Jumble is built on [Nostr][nostr], where your identity is tied to a private
digital key that only you hold. Small servers called [relays][relays], run by
people around the world, pass messages along without owning the account behind
them. No company can silence your account because no company owns it, or Nostr
itself.

> Other contenders include [Amethyst][amethyst], [Coracle][coracle], and

Reviewer: "Other contenders" reads as if those are contenders to nostr, since it's all about nostr in the previous paragraph.

> Other contenders include [Amethyst][amethyst], [Coracle][coracle], and

Author: ```suggestion
Other clients include [Amethyst][amethyst], [Coracle][coracle], and
```

what has driven him ever since.

Like most children in China, his life was structured around a single variable:
examinations. School, exams, more school, more exams, all the way to university,
and none of it suited him. "I was never the kind of student who followed rules
very well," he says. "I disliked exams and memorizing things that felt
meaningless to me." The consequence was a university he hadn't aimed for and a
major that wasn't computer science. Rules that felt arbitrary always chafed him.
So he follows his own code.

Outside of those structures, he moves at a different pace. On weekends he walks
with his girlfriend, to the sea, through parks, into bookstores, to restaurants
they haven't tried before. He doesn't plan far ahead. The four characters on his
GitHub profile are, he says, a reminder not to overthink, to act on what feels
right inside.

He doesn't remember exactly when he first encountered Bitcoin, only what made
him stay.

> Like many people, I dislike being controlled and want to truly own certain
> things. I think many Nostr users feel similarly. They want to truly own their
> data instead of having it locked inside the databases of large companies. 

> Bitcoin has a four-word phrase for this conviction: your keys, your coins. An

Reviewer: I was expecting to read "not your keys, not your coins" as a follow-up to that phrase.

> Bitcoin has a four-word phrase for this conviction: your keys, your coins. An

Author: ```suggestion
Bitcoin has a six-word phrase for this conviction: not your keys, not your coins. An
```

  darkSrc="/static/images/spotlight/cody-tseng/jumble-desktop-dark.png"
  alt="The Jumble desktop client"
  width={2240}
  height={2112}
  className="w-full h-auto"
/>

How to attract the people who would build and tend these communities is a
question nobody has answered yet. It is not obvious who should do the heavy
lifting, or why. His intuition is that the answer involves learning from the
platforms Nostr exists as an alternative to, studying how the big platforms keep
groups alive, and borrowing what actually works.

> I think the meaning of Nostr is that it provides another option. It shows
> people that another kind of network is possible.

If Nostr's promise is control over one's digital life, the test is whether
ordinary users can feel that control without first learning how the machine
works. Cody cannot resolve the larger question of whether Nostr will reach
beyond those who already know why it matters. But what he is building, steadily
and in the open, may help the answer emerge in its own time.

He doesn't talk like a rebel, but his work tells a different story. After all,
Jumble follows an older cypherpunk tradition: don't ask permission, write code.

> He chafed exams he found meaningless, imagined a social network nobody owned

Reviewer: > He chafed exams

I'm not familiar with this phrase / expression?

Author: To "chafe at" (or chafe under/against) is to feel annoyed, impatient, or restricted by a rule, boundary, or situation.

> He chafed exams he found meaningless, imagined a social network nobody owned

Author: ```suggestion
He chafed at exams he found meaningless, imagined a social network nobody owned
```

before he could build one, and added a feature to a project almost as an
afterthought, and watched it become the thing itself. Each time, he follows his
heart, not knowing where it will lead.

> 随心而动.

---

*Cody is a recipient of an [OpenSats grant][OpenSats grant], awarded in November 2025, supporting
one year of [Jumble][jumble-social]'s development. “OpenSats support allows me to
temporarily stop worrying so much about income or about boring career things like
promotions at work,” he says. “It allows me to spend more time and energy on
personal projects.”* 

*If you want to support Cody, recommend [Jumble][jumble-social] to your friends
and join the new social web. Our support for Cody’s work was made possible thanks

> to your generous donations to the [General Fund]. For comments, corrections, or

Reviewer: Nostr fund, not general fund. 

I'll fix it.

  - "The design of Nostr is extremely simple. You can describe its core principles in just a few sentences."
  - "Personally, I think improving the Nostr ecosystem itself is ultimately more important."
  - "I think the meaning of Nostr is that it provides another option. It shows people that another kind of network is possible."
  - "随心而动 — Follow your heart."
---

SHENZHEN, China — Cody Tseng is folded into the evening rush the way everyone
else is in this city, absorbed by a current of twenty million people, one more
developer heading home through streets that never fully dim. The metropolis,
often styled as China's Silicon Valley, rose from rice paddies in a single
lifetime, and the pace shows: most of its residents work for somebody else's
company. 

So does Cody. He writes backend code at a small firm. Nothing too special, he
says. Just a pretty ordinary path.

Then he goes home, rests for a while, and connects his laptop to an external
monitor. On that screen, his project has been branching without a master plan,
crossing into countries he has never visited, [picked up by][x21social]
strangers who [forked his code][fevela] and [built their own

> versions][jumble-ouro]. He browses those versions himself, mining them for ideas

Reviewer 2: [Here](https://github.com/OpenSats/website/pull/791/changes#r3336147051) I changed link to make it consisten with the other two, since they both link to the actual client, while `ouro` links to repo.

Also, `ouro` is a fork by Cody himself, just updated by an AI agent automatically. Don't know if it makes sense to add it inside the sentence `picked up by strangers who forked his code and built their own
versions`. What do you think?

Reviewer: We could just link to https://github.com/CodyTseng/jumble#forks

> versions][jumble-ouro]. He browses those versions himself, mining them for ideas

Reviewer 2: ```suggestion
crossing into countries he has never visited, picked up by
strangers who [forked his code][forks] and built their own
versions. He browses those versions himself, mining them for ideas
```

are willing to see, where they want to gather and whom they choose to trust.

That said, the "Great" Firewall complicates things in ways that outsiders rarely
map precisely.

Cody is careful not to overdramatise the Chinese internet. "The Chinese internet
itself is actually very diverse," he says. "People share daily life, jokes,
technology, art, and many other kinds of content. Chinese internet users are
also generally very friendly."

The system that sits behind the Chinese internet runs on different rules. Posts
can be deleted on a whim, keywords are routinely filtered, and the companies
hosting your words also keeps meticulous watch over everyone. Orwell would
recognise the architecture. Nostr is harder to pin down, but not beyond reach.
Cody says flatly that "using a VPN is basically required for development work",
and "without a VPN, using Nostr is more limited."

The Firewall does not block every unknown relay by default. It blocks the ones
that have been noticed, which means that, for now, obscurity is its own form of
protection. "Once you are using a VPN," he says, "where you are physically
located starts to matter a lot less." He laughs, as if the obstacle is real but

> not where the real difficulty lies.

Reviewer 2: `as if the obstacle is real but
not where the real difficulty lies.`

I get the meaning of this sentence, but it seems a bit off to me. Can you do a slight rephrase?

He doesn't talk like a rebel, but his work tells a different story. After all,
Jumble follows an older cypherpunk tradition: don't ask permission, write code.
He chafed at exams he found meaningless, imagined a social network nobody owned
before he could build one, and added a feature to a project almost as an
afterthought, and watched it become the thing itself. Each time, he follows his
heart, not knowing where it will lead.

> 随心而动.

---

Cody has been an OpenSats grantee since [June 2025][first-grant], with his grant
renewed in [November 2025][second-grant]. "OpenSats support allows me to
temporarily stop worrying so much about income or about boring career things
like promotions at work," he says. "It allows me to spend more time and energy
on personal projects."

If you want to support Cody, recommend [Jumble][jumble-social] to your friends
and join the new social web. Our support for Cody's work was made possible
thanks to your generous donations to [The Nostr Fund]. For comments,
corrections, or suggestions about our [Spotlight][OpenSats spotlight] series,
please reach out to spotlight@opensats.org.

[x21social]: https://x21.social/
[fevela]: https://fevela.me/

> [jumble-ouro]: https://github.com/CodyTseng/jumble-ouro

Reviewer 2: ```suggestion
[jumble-ouro]: https://ouro.jumble.social/
```

> [jumble-ouro]: https://ouro.jumble.social/

Reviewer 2: ```suggestion
[forks]: https://github.com/CodyTseng/jumble#forks
```
